common/progress.go

```diff
@@ -0,0 +1,61 @@
+package common
+
+import (
+	"fmt"
+	"github.com/cheggaaa/pb"
+	"github.com/hashicorp/packer/packer"
+	"log"
+	"time"
+)
+
+// Default progress bar appearance
+func GetDefaultProgressBar(ui *packer.Ui) pb.ProgressBar {
+	bar := pb.New64(0)
+	bar.ShowPercent = true
+	bar.ShowCounters = true
+	bar.ShowSpeed = false
+	bar.ShowBar = true
+	bar.ShowTimeLeft = false
+	bar.ShowFinalTime = false
+	bar.SetUnits(pb.U_BYTES)
+	bar.Format("[=>-]")
+	bar.SetRefreshRate(1 * time.Second)
+	bar.SetWidth(80)
+
+	// If there's no UI set, then the progress bar doesn't need anything else
+	if ui == nil {
+		return *bar
+	}
+	UI := *ui
+
+	// Now check the UI's width to adjust the progress bar
+	uiWidth := UI.GetMinimumLength() + len("\n")
```

Comment:
We could, but we'd actually still need to know what each UI instance prefixes messages with (such as for the `TargetedUI` where the `target` string is prefixed to each message) as otherwise it'll result in there being extra characters and thus an extra line between each render of the progress bar.

I couldn't figure out any other elegant way to determine what `UI.Message` prefixes each message with, so that's where the need for `UI.GetMinimumLength()` came from.

Reply:
Ah, I see. Tricky. Let me think on this for a few.

Reply:
Okay, so. I tinkered a bit with your tool, and one limitation of what you've done here is that it only works if PACKER_LOG=0. If PACKER_LOG=1, the "prefix" being added by the ui is larger, and it causes the progress bar to wrap still.  Which got me thinking. A true progress bar should be overwriting itself anyway.  I tinkered around with ANSI escape sequences and found that adding a "prefix" and "postfix" to the bar could allow us to overwrite the bar in place, which can also spare us the issue caused by the UI becasue we can overwrite it too.  Example code: 

```
	bar.SetWidth(width) // this is just the width gotten by GetTerminalDimensions() without any extra processing
	UI.Say("\n")
	// move up a line, erase it, and overwrite it.
	bar.Prefix("\033[K" + "\033[1A" + "\r")
	// drop down a line, clear it, then move back up to original line.
	bar.Postfix("\033[1B" + "\r" + "\033[K" + "\033[1A")
```

Output: 
(with logs enabled)
<img width="940" alt="screen shot 2018-08-09 at 1 04 38 pm" src="https://user-images.githubusercontent.com/1008838/43922857-d3807630-9bd4-11e8-8eff-afb0f83d97ac.png">

 (with logs disabled)
<img width="895" alt="screen shot 2018-08-09 at 1 05 26 pm" src="https://user-images.githubusercontent.com/1008838/43922902-f0d17176-9bd4-11e8-8ac1-0a77c033e06d.png">

Reply:
This may not work if we have parallel builds downloading multiple files though... Hmm.

Reply:
Yeah. I definitely prefer it to overwrite itself but leaned towards not doing it because of the use case where someone might want to redirect things to a file or maybe some other ui in the future that has some special case.

Maybe we can disable it for some of the UIs? Is that the only way we can determine if logging or something else is hooking the output?

Reply:
Also, since PACKER_LOG is an environment variable. We can check for its existence and skip drawing the progress bar in that situation too.